cli/cli-client/src/main/java/org/infinispan/cli/commands/Backup.java

```diff
@@ -0,0 +1,221 @@
+   public static class Get extends AbstractBackupCommand {
+      public static final String CMD = "get";
+      public static final String NO_CONTENT = "no-content";
+
+      @Argument(description = "The name of the backup", completer = BackupCompleter.class, required = true)
+      String name;
+
+      @Option(description = "No content is downloaded, but the command only returns once the backup has finished", hasValue = false, name = NO_CONTENT)
+      boolean noContent;
+
+      @Override
+      protected CommandResult exec(ContextAwareCommandInvocation invocation) {
+         return invocation.execute(
+               new CommandInputLine(Backup.CMD)
+                     .arg(TYPE, CMD)
+                     .arg(NAME, name)
+                     .option(NO_CONTENT, noContent)
+         );
+      }
+   }
+
+   @CommandDefinition(name = ListBackups.CMD, description = "List all backups on the server", activator = ConnectionActivator.class)
+   public static class ListBackups extends AbstractBackupCommand {
+      public static final String CMD = "ls";
+
+      @Override
+      protected CommandResult exec(ContextAwareCommandInvocation invocation) {
+         return invocation.execute(
+               new CommandInputLine(Backup.CMD)
+                     .arg(TYPE, CMD)
+         );
+      }
+   }
+
+   @CommandDefinition(name = Create.CMD, description = "Create a backup on the server", activator = ConnectionActivator.class)
+   public static class Create extends AbstractResourceCommand {
+      public static final String CMD = "create";
+      public static final String DIR = "dir";
+
+      @Option(shortName = 'd', description = "The directory on the server to be used for creating and storing the backup")
+      String dir;
+
+      @Option(shortName = 'n', description = "The name of the backup")
+      String name;
+
+      public Create() {
+         super(CMD);
+      }
+
+      @Override
+      protected void additionalArgs(CommandInputLine cmd) {
+         cmd.option(DIR, dir)
+               .option(NAME, name);
+      }
+   }
+
+   @CommandDefinition(name = Restore.CMD, description = "Restore a backup", activator = ConnectionActivator.class)
+   public static class Restore extends AbstractResourceCommand {
+      public static final String CMD = "restore";
+      public static final String UPLOAD_BACKUP = "upload-backup";
+
+      @Argument(description = "The path of the backup file ", completer = FileOptionCompleter.class, required = true)
+      Resource path;
+
+      @Option(shortName = 'u', description = "Indicates that the path is a local file which must be uploaded to the server", hasValue = false, name = UPLOAD_BACKUP)
+      boolean upload;
+
+      public Restore() {
+         super(CMD);
+      }
+
+      @Override
+      protected void additionalArgs(CommandInputLine cmd) {
+         if (!upload && !((FileResource) path).getFile().isAbsolute())
```

Comment:
the cast can't fail ?

Reply:
I don't think so, the only other implementation is `PipelineResource` which returns `""` or `false` for all methods. Unfortunately it's not possible to use `FileResource` directly as the variable type, because aesh does not know how to set the value, resulting in a NPE being thrown.